### PR TITLE
IPC via abstract unix domain sockets

### PR DESCRIPTION
**Pull request by serges147**
- Added new "SDK" library target; added `ocvsmd-cli` executable target to consume the SDK.
- Posix calls now are wrapped with `platform::posixSyscallError` (which does retries on `EINTR`). For now only IPC stuff correctly wrapped; the same for other places (like executors, udp sockets, etc.) to be continued...

Also:
- Now compiles on macOS by adding kqueue-based implementation of executor for Darwin. Thank you @thirtytwobits for sharing your original implementation (I did some fixes related to `kevent` timeout calculations - now it's **nano**seconds based instead of seconds).

### CODE REVIEW

File: src/daemon/main.cpp [hunk 1]
```diff
@@ -66,7 +66,7 @@ void step_01_close_all_file_descriptors(std::array<int, 2>& pipe_fds)
         ::exit(EXIT_FAILURE);
     }
     constexpr int first_fd_to_close = 3;  // 0, 1 & 2 for standard input, output, and error.
-    for (int fd = first_fd_to_close; fd <= rlimit_files.rlim_max; ++fd)
+    for (int fd = first_fd_to_close; fd <= rlimit_files.rlim_cur; ++fd)
```
Comment (serges147) [inline src/daemon/main.cpp:69]: On macOS `rlim_max` is huge!!!

File: src/common/ipc/unix_socket_client.hpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    UnixSocketClient(UnixSocketClient&&)                 = delete;
+    UnixSocketClient(const UnixSocketClient&)            = delete;
+    UnixSocketClient& operator=(UnixSocketClient&&)      = delete;
+    UnixSocketClient& operator=(const UnixSocketClient&) = delete;
+
+    ~UnixSocketClient();
+
+    bool connect_to_server();
+    void send_message(const std::string& message) const;
```
Comment (thirtytwobits) [inline src/common/ipc/unix_socket_client.hpp:31]: can we also support string_view?

File: src/common/ipc/unix_socket_server.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    }
+
+    sockaddr_un addr{};
+    addr.sun_family                        = AF_UNIX;
+    const std::string abstract_socket_path = '\0' + socket_path_;
+    CETL_DEBUG_ASSERT(abstract_socket_path.size() <= sizeof(addr.sun_path), "");
+    // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-array-to-pointer-decay,hicpp-no-array-decay)
+    std::memcpy(addr.sun_path,
+                abstract_socket_path.c_str(),
+                std::min(sizeof(addr.sun_path), abstract_socket_path.size()));
```
Comment (serges147) [inline src/common/ipc/unix_socket_server.cpp:123]: On BSD I'm gonna return back to usual Unix domain sockets - BSD doesn't support abstract ones.

File: src/common/ipc/unix_socket_server.hpp [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    UnixSocketServer& operator=(const UnixSocketServer&) = delete;
+
+    ~UnixSocketServer();
+
+    bool start();
+
+    CETL_NODISCARD libcyphal::IExecutor::Callback::Any registerListenCallback(
+        libcyphal::IExecutor::Callback::Function&& function) const;
+
+    void accept();
```
Comment (thirtytwobits) [inline src/common/ipc/unix_socket_server.hpp:60]: Shouldn't this return an error?

File: src/common/ipc/unix_socket_server.cpp (hunk 3 above)
Comment (serges147) [reply src/common/ipc/unix_socket_server.cpp:123]: BTW, @thirtytwobits let me know please, do you seek that our ocvsmd could run on macOS?
For now it is not a problem (besides the above comment about abstract unix domain sockets), but I would like to know whether it worth to spend a bit extra efforts to this or not. Thnx

Comment (thirtytwobits) [reply src/common/ipc/unix_socket_server.cpp:123]: It's not required for the contract work, no. It's always helpful to have for debugging.

File: src/common/ipc/unix_socket_server.hpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+private:
+    void handle_client_connection(const int client_fd);
+    void handle_client_request(const int client_fd);
+
+    libcyphal::IExecutor&                                           executor_;
+    const std::string                                               socket_path_;
+    int                                                             server_fd_;
+    platform::IPosixExecutorExtension* const                        posix_executor_ext_;
+    std::unordered_map<int, std::unique_ptr<detail::ClientContext>> client_contexts_;
```
Comment (thirtytwobits) [inline src/common/ipc/unix_socket_server.hpp:70]: nit: might be more elegant to use a `std::set<std::unique_ptr<detail::ClientContext>>` where `detail::ClientContext` had a `int GetHandle() const` method and you provided a custom comparitor that used that.

File: src/common/ipc/unix_socket_client.hpp (hunk 2 above)
Comment (serges147) [reply src/common/ipc/unix_socket_client.hpp:31]: this `send_message` was just to prove send and receive. real communication api of course will be different. probably passing back and forth some serialized structures.

File: src/common/ipc/unix_socket_server.hpp (hunk 4 above)
Comment (serges147) [reply src/common/ipc/unix_socket_server.hpp:60]: Normally it should but in my case there nowhere to return it actually. See call site:
```
    ipc_server_callback_ = ipc_server_.registerListenCallback([this](const auto&) {
        //
        ipc_server_.accept();
    });
```

**Review by thirtytwobits — APPROVED** [review]

File: CMakeLists.txt [hunk 6]
```diff
@@ -67,6 +67,13 @@ add_definitions(
         -DVCS_REVISION_ID=0x${vcs_revision_id}ULL
         -DNODE_NAME="org.opencyphal.ocvsmd"
 )
+if (DEFINED PLATFORM_LINUX_TYPE)
+    if(${PLATFORM_LINUX_TYPE} STREQUAL "bsd")
+        add_definitions(-DPLATFORM_LINUX_TYPE_BSD)
```
Comment (pavel-kirienko) [inline CMakeLists.txt:72]: BSD is not Linux, it's a different kernel family, so `PLATFORM_LINUX_TYPE` may need renaming.

File: CMakeLists.txt [hunk 7]
```diff
@@ -67,6 +67,13 @@ add_definitions(
...
+    elseif(${PLATFORM_LINUX_TYPE} STREQUAL "debian")
+        add_definitions(-DPLATFORM_LINUX_TYPE_DEBIAN)
```
Comment (pavel-kirienko) [inline CMakeLists.txt:74]: Debian comes with BSD kernels as well, not just Linux. While we are at it, Linux is a kernel, not an OS, that would be GNU/Linux. Should we not simply go like `PLATFORM_DEBIAN`, `PLATFORM_BSD`? Although then it seems odd that one of them is an OS while the other is a kernel.

**Review by pavel-kirienko — APPROVED** [review]

**Pull request merged**